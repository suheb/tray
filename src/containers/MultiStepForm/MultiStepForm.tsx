import React, { useState, useEffect } from "react";
import { Grid, Button, ButtonGroup } from "@material-ui/core";
import UserForm from "../UserForm/UserForm";
import PrivacyForm from "../PrivacyForm/PrivacyForm";
import DoneForm from "../DoneForm/DoneForm";


const MultiStepForm = () => {
  const [activeStep, setActiveStep] = useState(0);
  const [fullFormData, setFullFormData] = useState<{ [name: string]: Object }>({});

  const handleSubmit = (formName: string, formData: { [name: string]: any }) => {
    setFullFormData({
      ...fullFormData,
      [formName]: formData
    });
    setActiveStep(activeStep + 1);
  }
  const handleBack = () => setActiveStep(activeStep - 1)
  const getStepContent = (step: number) => Object.values(steps)[step]
  const isActiveStep = (step: number) => step === activeStep

  const steps = {
    "User": (
      <UserForm
        formData={fullFormData["user"]}
        handleSubmit={handleSubmit}
      />
    ),
    "Privacy": (
      <PrivacyForm
        formData={fullFormData["privacy"]}
        handleSubmit={handleSubmit}
        handleBack={handleBack} />
    ),
    "Done": (
      <DoneForm />
    )
  }

  useEffect(() => {
    if (activeStep === Object.keys(steps).length - 1) {
      console.log(fullFormData);
    }
  }, [activeStep, steps, fullFormData]);

  return (
    <Grid
      container
      direction="column"
      alignContent="center"
    >
      <Grid
        item
        xs={4}
      >
        <ButtonGroup color="secondary" fullWidth>
          {Object.keys(steps).map((step, index) => <Button key={step} disabled={!isActiveStep(index)}>{step}</Button>)}
        </ButtonGroup>
      </Grid>
      <Grid
        item
        xs={4}
      >
        <>
          {getStepContent(activeStep)}
        </>
      </Grid>
    </Grid>
  )
}

export default MultiStepForm;
