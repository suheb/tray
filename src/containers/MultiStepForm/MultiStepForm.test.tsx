import React from 'react';
import { render } from '@testing-library/react';

import MultiStepForm from './MultiStepForm';
import userEvent from '@testing-library/user-event';

describe("MultiStepForm", () => {
  it("renders without crashing", () => {
    const { asFragment } = render(<MultiStepForm />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders the first page by default", () => {
    const { getByText, getByLabelText } = render(<MultiStepForm />);
    expect(getByText(/user/i)).not.toBeDisabled();
    expect(getByLabelText(/name/i)).toBeInTheDocument();
  });

  it("renders the second page by successful submit of first page", async () => {
    const { getByText, getByLabelText } = render(<MultiStepForm />);

    await userEvent.type(getByLabelText(/name/i), "John Smith")
    await userEvent.type(getByLabelText(/role/i), "Analyst")
    await userEvent.type(getByLabelText(/email/i), "abc@gmail.com")
    await userEvent.type(getByLabelText(/password/i), "1Abcdefdag")

    userEvent.click(getByText(/submit/i));
    expect(getByText(/privacy/i)).not.toBeDisabled();
    expect(getByLabelText(/receive updates/i)).toBeInTheDocument();
  });
  
  it("renders the first page by clicking back on second page with correct data", async () => {
    const { getByText, getByLabelText } = render(<MultiStepForm />);

    await userEvent.type(getByLabelText(/name/i), "John Smith")
    await userEvent.type(getByLabelText(/role/i), "Analyst")
    await userEvent.type(getByLabelText(/email/i), "abc@gmail.com")
    await userEvent.type(getByLabelText(/password/i), "1Abcdefdag")

    userEvent.click(getByText(/submit/i));
    expect(getByText(/privacy/i)).not.toBeDisabled();
    expect(getByLabelText(/receive updates/i)).toBeInTheDocument();
    userEvent.click(getByText(/back/i));
    expect(getByText(/user/i)).not.toBeDisabled();
    expect((getByLabelText(/name/i) as HTMLInputElement).value).toBe("John Smith");
  });

});
