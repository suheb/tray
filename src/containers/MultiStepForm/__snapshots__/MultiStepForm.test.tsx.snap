// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiStepForm renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-align-content-xs-center"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
    >
      <div
        class="MuiButtonGroup-root MuiButtonGroup-fullWidth"
        role="group"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedSecondary MuiButton-outlinedSecondary MuiButton-fullWidth"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            User
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedSecondary MuiButton-outlinedSecondary Mui-disabled MuiButton-fullWidth Mui-disabled"
          disabled=""
          tabindex="-1"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Privacy
          </span>
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedSecondary MuiButton-outlinedSecondary Mui-disabled MuiButton-fullWidth Mui-disabled"
          disabled=""
          tabindex="-1"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Done
          </span>
        </button>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-justify-xs-flex-end"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
        >
          <div
            class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated Mui-required Mui-required"
              data-shrink="false"
              for="name"
              id="name-label"
            >
              Name
              <span
                class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
              >
                 *
              </span>
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                id="name"
                required=""
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
        >
          <div
            class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
              data-shrink="false"
              for="role"
              id="role-label"
            >
              Role
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                id="role"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
        >
          <div
            class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated Mui-required Mui-required"
              data-shrink="false"
              for="email"
              id="email-label"
            >
              Email
              <span
                class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
              >
                 *
              </span>
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                id="email"
                required=""
                type="email"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
        >
          <div
            class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated Mui-required Mui-required"
              data-shrink="false"
              for="password"
              id="password-label"
            >
              Password
              <span
                class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
              >
                 *
              </span>
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                id="password"
                required=""
                type="password"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-fullWidth"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Submit
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
